### ISSUE TITLE
Creating aws_eks_cluster in the same module where Kubernetes provider resources are also used  

### ISSUE THREAD

**Issue by jennerm** [terraform-aws-modules/terraform-aws-eks#2784]
## Description

I am creating some terraform modules to create our EKS cluster and then install a few components using the hashicorp/kubernetes hashicorp/helm providers.

The documentation for both providers has clear warnings that the cluster should be created in a separate apply from the resources being created by the Kubernetes and Helm providers e.g: From Kubernetes provider [https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs ](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs)

"When using interpolation to pass credentials to the Kubernetes provider from other resources, these resources SHOULD NOT be created in the same Terraform module where Kubernetes provider resources are also used."

However, when I look at this EKS blueprint I see them being mixed.  The cluster is created in main.tf on line 25:
https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/main.tf#L25

And the aws-auth configmap is created in the same file on line 537 using a Kubernetes provider kubernetes_config_map resource:
https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/main.tf#L537

And when the module is used, the provider is configured using credentials interpolated from the cluster resource:
https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/eks_managed_node_group/main.tf#L5

This seems to directly contradict the advice given in the Kubernetes provider.

When I tried using a single module, the create was successful several times, but when I updated some resources and ran apply operations a few hours later I received an error message like this:
```
│ Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable

│ Error: Get "http://localhost/apis/scheduling.k8s.io/v1/priorityclasses/unspecified-priority": dial tcp [::1]:80: connect: connection refused
```
Which I can see reported in other places such as https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1234 and the ultimate recommendation seems to be to split the modules.

Can you please tell me:
1. Why does a single module work here? Is there a particular sequence of dependencies that I must replicate?

2. Is it safe to copy this for creating and updating a production cluster or is this just for the simplicity of creating a cluster for this example. Would it be safer for me to have two separate modules, one for the cluster and another for the components as advised by the Kubernetes Provider?

## Versions
- Terraform version:
Terraform v1.5.7
- Provider version(s):
-  hashicorp/helm v2.11.0
-  hashicorp/kubernetes v2.23.0
-  hashicorp/aws v5.1.0
-  hashicorp/random v3.5.1
-  hashicorp/tls v3.4.0
-  hashicorp/null v3.2.1
-  hashicorp/template v2.2.0
-  hashicorp/local v2.4.0

**Comment by rwd5213** [comment]
Given #2541 and what you said about them giving warnings about doing this, it might be a better idea to separate them out into a different module. I will probably go that path in creating a simple other module to handle the auth configmap

**Comment by bryantbiggs** [comment]
> simple other module to handle the auth configmap

I would recommend not doing this - its a sure fire way of cutting off access for your nodes.

**Comment by jennerm** [comment]
Thanks for the quick response. I have decided to go with two modules, one to create the aws_eks_cluster and various other bootstrap resources, including the node groups.  And a second module for everything else, with clear dependencies on the aws-auth configmap so that it is created first and destroyed last (and a big comment about its importance to the nodes).

Now that I understand better the distinction between the EKS Blueprints and this repo, it also seems that they are moving towards multiple modules too, using this to create the cluster and then adding the rest of the resources separately.  I found a note about keeping the cluster separate in the Terraform caveats section in their documentation too: https://aws-ia.github.io/terraform-aws-eks-blueprints/
